File: stream-comonad.ts
export class Stream<A> {
  constructor(
    public readonly head: A,
    public readonly tail: () => Stream<A>,
  ) { }

  map<B>(f: (a: A) => B): Stream<B> {
    return new Stream(f(this.head), () => this.tail().map(f));
  }

  extract(): A {
    return this.head;
  }

  duplicate(): Stream<Stream<A>> {
    return new Stream(this, () => this.tail().duplicate());
  }

  extend<B>(f: (wa: Stream<A>) => B): Stream<B> {
    return this.duplicate().map(f);
  }

  take(n: number): A[] {
    const result: A[] = [];
    let current: Stream<A> = this;
    for (let i = 0; i < n; i++) {
      result.push(current.head);
      current = current.tail();
    }
    return result;
  }
}

export const from = (n: number): Stream<number> =>
  new Stream(n, () => from(n + 1));

export const sumS = (n: number, s: Stream<number>): number => {
  let sum = 0;
  let current = s;
  for (let i = 0; i < n; i++) {
    sum += current.head;
    current = current.tail();
  }
  return sum;
};

export const average = (n: number, s: Stream<number>): number => {
  return sumS(n, s) / n;
};

export const movingAvg = (n: number, s: Stream<number>): Stream<number> =>
  s.extend((w) => average(n, w));

const s = from(1);
const avg = movingAvg(3, s);
console.log(avg.take(10));

File: stream-comonad.hs
{-# LANGUAGE InstanceSigs #-}
module Main where

class Functor w => Comonad w where
  extract :: w a -> a
  duplicate :: w a -> w (w a)
  extend :: (w a -> b) -> w a -> w b
  extend f = fmap f . duplicate

data Stream a = Cons a (Stream a)

instance Functor Stream where
  fmap :: (a -> b) -> Stream a -> Stream b
  fmap f (Cons a as) = Cons (f a) (fmap f as)

instance Comonad Stream where
  extract (Cons a _) = a

  duplicate (Cons a as) = Cons (Cons a as) (duplicate as)
  
tail :: Stream a -> Stream a
tail (Cons a as) = as

sumS :: Num a => Int -> Stream a -> a
sumS n (Cons a as) = if n <= 0 then 0 else a + sumS (n - 1) as

average :: Fractional a => Int -> Stream a -> a
average n stm = (sumS n stm) / (fromIntegral n)

movingAvg :: Fractional b => Int -> Stream b -> Stream b
movingAvg n = extend (average n)

main :: IO ()
main = putStrLn "Hello"